### PR TITLE
Fix: area chart stacking doesn't work

### PR DESCRIPTION
**Pull request by machira**
This PR addresses issue #948 as recommended by Plotly maintainers [here](https://github.com/plotly/plotly.js/issues/344). 

Specifically, we recalculate the data appearance when the legend is clicked. This ensures that we show existing series at the updated heights. 

We also fix the behaviour of percentage area plots. Here, if the legend is clicked to disable a given series, the visualization is refreshed so that only enabled series are plotted. The percentage total is still calculated over the all series. This is designed to be consistent with the percentage bar charts, where after disabling a given series, the plotted visualizations add up to less than 100%.

### CODE REVIEW

File: rd_ui/app/scripts/directives/plotly.js [hunk 1]
```diff
@@ -45,59 +45,82 @@
...
+    _.each(seriesList, function(series) {
+      if(!_.has(series,'visible')){
+        series.visible = true;
+        series.original_y = series.y.slice();
+      }
+    });
+  };
+
+  var getEnabledSeries = function(seriesList){
+    return _.filter(seriesList, function(series) {
```
Comment (arikfr) [inline rd_ui/app/scripts/directives/plotly.js]: You can write this as `_.filter(seriesList, 'visible');`.

**Comment by arikfr** [comment]
Overall looks good. Did you test it with a chart having multiple types of series (bars & line, for example)?

File: rd_ui/app/scripts/directives/plotly.js [hunk 2]
```diff
@@ -288,6 +311,25 @@
           var element = element[0].children[0];
           Plotly.newPlot(element, scope.data, scope.layout, scope.plotlyOptions);
 
+          element.on('plotly_afterplot', function(d) {
+            if(scope.options.globalSeriesType === 'area' && (scope.options.series.stacking === 'normal' || scope.options.series.stacking === 'percent')){
+              d3.selectAll('.legendtoggle').each(function(d, i) {
+                // override plotly's default legend item on-click behaviour
```
Comment (arikfr) [inline rd_ui/app/scripts/directives/plotly.js]: What's the default behavior? Do we "un-override" it when the types changes?

File: rd_ui/app/scripts/directives/plotly.js (hunk 1 above)
Comment (machira) [inline rd_ui/app/scripts/directives/plotly.js]: Actually, this would not be the same. Because of Javascript truthiness, when `series.visible = 'legendonly'` would evaluate to `true`.

File: rd_ui/app/scripts/directives/plotly.js (hunk 2 above)
Comment (machira) [inline rd_ui/app/scripts/directives/plotly.js]: Thank you for the suggestion. We went back and took a look at it. It turns out we can use a custom listener through d3, and remove it when chart type changes.

**Comment by machira** [comment]
It looks like multiple series (bar and Area) breaks the stacking. I will keep looking at it, and update you if I find a workaround.

File: rd_ui/app/scripts/directives/plotly.js (hunk 1 above)
Comment (arikfr) [inline rd_ui/app/scripts/directives/plotly.js]: :+1:

**Comment by arikfr** [comment]
@machira ok, let me know if you find a solution. We can merge as is if you prefer, because it's still better than current situation.

**Comment by vorakumar** [comment]
Hey Arik, we have updated the PR with some new changes. Some additional tests here showed some edge cases we hadn't thought of- such as edit mode. The most recent push works as expected. And multiple series (area and line) also seem to work appropriately.

File: rd_ui/app/scripts/directives/plotly.js [hunk 3]
```diff
@@ -121,7 +144,6 @@
   angular.module('plotly', [])
     .constant('ColorPalette', ColorPalette)
     .directive('plotlyChart', function () {
-      var bottomMargin = 50;
```
Comment (arikfr) [inline rd_ui/app/scripts/directives/plotly.js]: Did you mean to remove this and the related code?

File: rd_ui/app/scripts/directives/plotly.js [hunk 4]
```diff
@@ -286,14 +304,31 @@
...
           scope.plotlyOptions = {showLink: false, displaylogo: false};
           scope.data = [];
 
           var element = element[0].children[0];
           Plotly.newPlot(element, scope.data, scope.layout, scope.plotlyOptions);
 
+          element.on('plotly_afterplot', function(d) {
+            if(scope.options.globalSeriesType === 'area' && (scope.options.series.stacking === 'normal' || scope.options.series.stacking === 'percent')){
+              $(element).find(".legendtoggle").each(function(i, rectDiv) {
+                  d3.select(rectDiv).on('click', function () {
```
Comment (arikfr) [inline rd_ui/app/scripts/directives/plotly.js]: Using `click` without a "namespace" mean we override the default?

**Comment by arikfr** [comment]
@vorakumar thanks! I appreciate the effort you put here.

File: rd_ui/app/scripts/directives/plotly.js (hunk 4 above)
Comment (vorakumar) [inline rd_ui/app/scripts/directives/plotly.js]: yes. We tried using `namespace` but it didn't work out. The new listener method was never called.

**Comment by vorakumar** [comment]
major work has been done by @machira

**Comment by arikfr** [comment]
By "you" I meant all of you at ThoughtWorks :-)

File: rd_ui/app/scripts/directives/plotly.js (hunk 4 above)
Comment (arikfr) [inline rd_ui/app/scripts/directives/plotly.js]: And is there a way to get the previous click handler and call it inside our own?

I'm worried that at some point in the future, Plotly will add some additional logic there that we will miss.

Comment (machira) [inline rd_ui/app/scripts/directives/plotly.js]: Yes, it is a risk that if Plotly updates that area, our stuff breaks. On the other hand, such an update may also fix the plotly issue, so we may want to remove this behaviour entirely. 
We tried a few things, but it didn't seem to work. Ultimately, we decided not to spend too much time on it since Plotly may be soon replaced by a different charting library. If it becomes an issue, we can revisit and rework this.

File: rd_ui/app/scripts/directives/plotly.js (hunk 3 above)
Comment (arikfr) [inline rd_ui/app/scripts/directives/plotly.js]: @machira @vorakumar ?

File: rd_ui/app/scripts/directives/plotly.js (hunk 4 above)
Comment (arikfr) [inline rd_ui/app/scripts/directives/plotly.js]: OK, sounds reasonable.

**Comment by rohanpd** [comment]
I have seen splunk do this in a different way that might be worth considering:

image

When you hover over the legend it just highlights the slice that you are analysing:

image

In saying that, I'm happy with what you guys have suggested.

File: rd_ui/app/scripts/directives/plotly.js (hunk 3 above)
Comment (machira) [inline rd_ui/app/scripts/directives/plotly.js]: @arikfr, No, we didn't mean to do that. It's now fixed.

**Comment by arikfr** [comment]
Merged :tada:
